chap-cheri-x86-64.tex

```diff
@@ -1260,3 +1260,300 @@ \subsection{XCHG [ER]AX Opcodes}
+and the distribution of the selector is guarded by capabilities.
+This could bypass the operating system for key management, allowing
+meaningful key management and distribution with virtual addresses
+directly from user space.
+
+This could also enable a light-weight CHERI enclave mechanism where
+the operating system would not have the ability to read the unencrypted
+contents of enclave memory despite managing the physical pages.
+In order to prevent keyed capabilities leaking to the OS, such a
+system would likely require storing keyed capabilities within the
+encrypted memory itself, and delegating register save and restore
+to trusted CPU state machines.
+
+Moving key management from physical pages to virtual addresses also
+allows ME to select keys on allocation granularity rather than on
+page granularity.
+When combined with capabilities, this could allow guarantees about
+memory sharing even where capability bounds overlap (by relying on
+encryption) and where keys are
+reused (by relying on capabilities).
+Allowing parts of a cache line to be encrypted with different keys
+could cause problems if we want to decrypt into the L1 cache, so the
+key selector would likely need to be treated like an extension of
+the physical address from the cache-tag perspective, with a distinct
+copy of the cache line held for each key.
+
+ME could also enable light-weight temporal safety for CHERI, where
+the encryption key for new allocations is cycled while leaving the
+memory unchanged.
+
+\subsection{Intel MPK}
+
+Intel's Memory Protection Keys (MPK) enables userspace filtering
+of read/write permissions of 15 sets of pages.
+The page table leaf entries themselves have a new 4-bit field to
+identify a page set (or key), with the value 0 wired to unfiltered.
+MPK was introduced with Skylake in 2015, and only appears to have
+been enabled for Xeon processors.
+AMD's Zen3 processors also appear to have MPK support since 2020,
+presumably enabled for both consumer and professional variants.
+
+In userspace, a thread may write its PKRU register, which maps
+16 pairs of read/write permissions coorosponding to the 16 possible
+keys.
+By writing this register, a thread may quickly remove access to subsets
+of pages.
+Current implementations appear to flush the back end of the pipeline,
+incuring a cycle penalty of roughly 20 cycles.
+This filtering is naturally per-thread, unlike the page table iteself
+which is per-process, so it can be used for fine-grained thread-local
+access for a few sensitive structures.
+For example, secret keys might be stored in a region of memory that
+is only briefly made accessible when necessary.
+
+MPK has 3 major disadvantages compared to CHERI capabilities:
+scaling, PKRU protection, and buffer sharing.
+
+While MPK appears to be extremely unscalable on the surface, work
+has been done to virtualise the page sets with an API similar to
+memprotect so that sets of pages can have their access permissions
```

Comment:
```suggestion
mprotect so that sets of pages can have their access permissions
```

Reply:
Not sure if you intend a specific memprotect() function distinct from the POSIX mprotect()?  If so, this might need a citation?